fdbserver/BlobMigrator.actor.cpp

```diff
@@ -59,22 +60,47 @@ static inline void dprint(fmt::format_string<T...> fmt, T&&... args) {
 // StorageServerInterface APIs which are needed for DataDistributor to start data migration.
 class BlobMigrator : public NonCopyable, public ReferenceCounted<BlobMigrator> {
 public:
-	BlobMigrator(Database db, BlobMigratorInterface interf) : interf_(interf), actors_(false), db_(db) {
+	BlobMigrator(Database db, BlobMigratorInterface interf, Reference<AsyncVar<ServerDBInfo> const> dbInfo)
+	  : interf_(interf), actors_(false), db_(db), dbInfo(dbInfo) {
 		blobConn_ = BlobConnectionProvider::newBlobConnectionProvider(SERVER_KNOBS->BLOB_RESTORE_MANIFEST_URL);
 	}
 	~BlobMigrator() {}
 
 	// Start migration
 	ACTOR static Future<Void> start(Reference<BlobMigrator> self) {
 		wait(checkIfReadyForMigration(self));
-		wait(lockDatabase(self->db_, self->interf_.id()));
+		wait(checkRangeEmptyAndLockDb(self, normalKeys));
 		wait(prepare(self, normalKeys));
 		wait(advanceVersion(self));
 		wait(serverLoop(self));
 		return Void();
 	}
 
 private:
+	// check the restored range is empty before lock the database, maybe in the future we can just lock the range.
+	ACTOR static Future<Void> checkRangeEmptyAndLockDb(Reference<BlobMigrator> self, KeyRangeRef keys) {
+		state Transaction tr(self->db_);
+		UID debugID = deterministicRandom()->randomUniqueID();
+		tr.debugTransaction(debugID);
+		loop {
+			try {
+				// sanity check there's no data in the restored overlapped range
+				RangeResult rangeResult = wait(tr.getRange(keys, 1));
+				if (!rangeResult.empty()) {
+					throw restore_destination_not_empty();
+				}
+				wait(lockDatabase(&tr, self->interf_.id()));
+				wait(tr.commit());
+				TraceEvent("BlobRestorePrepare", self->interf_.id()).detail("State", "LockDatabase");
```

Comment:
nit - can you print the lock id here? in some cases we may need to use this id to unlock db manually.

Reply:
The lock id is the same as the interface id.